homeassistant/block_async_io.py

```diff
@@ -46,53 +48,139 @@ def _check_sleep_call_allowed(mapped_args: dict[str, Any]) -> bool:
+
+    calls: set[BlockingCall]
+
+
+_BLOCKED_CALLS = BlockedCalls(set())
+
+
 def enable() -> None:
     """Enable the detection of blocking calls in the event loop."""
-    loop_thread_id = threading.get_ident()
-    # Prevent urllib3 and requests doing I/O in event loop
-    HTTPConnection.putrequest = protect_loop(  # type: ignore[method-assign]
-        HTTPConnection.putrequest, loop_thread_id=loop_thread_id
-    )
-
-    # Prevent sleeping in event loop.
-    time.sleep = protect_loop(
-        time.sleep,
-        check_allowed=_check_sleep_call_allowed,
-        loop_thread_id=loop_thread_id,
-    )
-
-    glob.glob = protect_loop(
-        glob.glob, strict_core=False, strict=False, loop_thread_id=loop_thread_id
-    )
-    glob.iglob = protect_loop(
-        glob.iglob, strict_core=False, strict=False, loop_thread_id=loop_thread_id
-    )
-    os.walk = protect_loop(
-        os.walk, strict_core=False, strict=False, loop_thread_id=loop_thread_id
-    )
-
-    if not _IN_TESTS:
-        # Prevent files being opened inside the event loop
-        os.listdir = protect_loop(  # type: ignore[assignment]
-            os.listdir, strict_core=False, strict=False, loop_thread_id=loop_thread_id
-        )
-        os.scandir = protect_loop(  # type: ignore[assignment]
-            os.scandir, strict_core=False, strict=False, loop_thread_id=loop_thread_id
-        )
+    if _BLOCKED_CALLS.calls:
+        raise RuntimeError("Blocking call detection is already enabled")
 
-        builtins.open = protect_loop(  # type: ignore[assignment]
-            builtins.open,
-            strict_core=False,
-            strict=False,
-            check_allowed=_check_file_allowed,
+    loop_thread_id = threading.get_ident()
+    for blocking_call in BLOCKING_CALLS:
+        if _IN_TESTS and blocking_call.skip_for_tests:
+            continue
+
+        protected_function = protect_loop(
+            blocking_call.original_func,
+            strict=blocking_call.strict,
+            strict_core=blocking_call.strict_core,
+            check_allowed=blocking_call.check_allowed,
             loop_thread_id=loop_thread_id,
         )
-        # unittest uses `importlib.import_module` to do mocking
-        # so we cannot protect it if we are running tests
-        importlib.import_module = protect_loop(
-            importlib.import_module,
-            strict_core=False,
-            strict=False,
-            check_allowed=_check_import_call_allowed,
-            loop_thread_id=loop_thread_id,
+        setattr(blocking_call.object, blocking_call.function, protected_function)
+        _BLOCKED_CALLS.calls.add(blocking_call)
+
+
+def disable() -> None:
+    """Disable the detection of blocking calls in the event loop."""
+    for blocking_call in _BLOCKED_CALLS.calls:
+        setattr(
+            blocking_call.object, blocking_call.function, blocking_call.original_func
         )
+    _BLOCKED_CALLS.calls.clear()
```

Comment:
We may want to move this to tests since it's only needed in tests

Reply:
9e231df83a61626221678dab225504d158d28d52